server/api/actions.go

```
@@ -0,0 +1,151 @@
+		return
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{
+		ID: id,
+	}
+	w.Header().Add("Location", makeAPIURL(a.pluginAPI, "actions/channel/%s/%s", channelAction.ChannelID, id))
+
+	ReturnJSON(w, &result, http.StatusCreated)
+}
+
+func (a *ActionsHandler) getChannelActions(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+
+	vars := mux.Vars(r)
+	channelID := vars["channel_id"]
+
+	if !a.PermissionsCheck(w, a.permissions.ChannelActionView(userID, channelID)) {
+		return
+	}
+
+	params := r.URL.Query()
+	triggerType := params.Get("trigger_type")
+
+	actions, err := a.channelActionsService.GetChannelActions(channelID, triggerType)
+	if err != nil {
+		a.HandleErrorWithCode(w, http.StatusInternalServerError, fmt.Sprintf("unable to retrieve actions for channel %s", channelID), err)
+		return
+	}
+
+	ReturnJSON(w, &actions, http.StatusOK)
+}
+
+// checkAndSendMessageOnJoin handles the GET /actions/channels/{channel_id}/check_and_send_message_on_join endpoint.
+func (a *ActionsHandler) checkAndSendMessageOnJoin(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	channelID := vars["channel_id"]
+	userID := r.Header.Get("Mattermost-User-ID")
+
+	if !a.PermissionsCheck(w, a.permissions.ChannelActionView(userID, channelID)) {
+		return
+	}
+
+	hasViewed := a.channelActionsService.CheckAndSendMessageOnJoin(userID, channelID)
+	ReturnJSON(w, map[string]interface{}{"viewed": hasViewed}, http.StatusOK)
+}
+
+func (a *ActionsHandler) updateChannelAction(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+
+	vars := mux.Vars(r)
+	channelID := vars["channel_id"]
+
+	if !a.PermissionsCheck(w, a.permissions.ChannelActionUpdate(userID, channelID)) {
+		return
+	}
+
+	var newChannelAction app.GenericChannelAction
+	if err := json.NewDecoder(r.Body).Decode(&newChannelAction); err != nil {
+		a.HandleErrorWithCode(w, http.StatusBadRequest, "unable to parse action", err)
+		return
+	}
+
```

The channelID in the body params of the PUT request might be different from the one specified in the Query Param. For example something like this:

```
PUT /channels/{channel_id}/{action_id}

{"channel_id": "different_channel_id".........}
```

In this case, the validation to check if user has write permissions to channel will pass. However when the server calls `a.channelActionsService.Update(newChannelAction)` it might actually be updating a different channel ID. 

Hence, a suggested fix is to include this:

```
	if newChannelAction.ChannelID != channelID {
		newChannelAction.ChannelID = channelID
	}
```

---

Instead of silently changing the body contents, I returned a 400 Bad Request Error, just to make the life of client developers easier: if the client has a bug where the URL and the body are different, I believe it's better to fail early and loudly.

---

@agarciamontoro can we make the behavior consistent in `createChannelAction` function as well. There we override and proceed further - like this:

```
	if channelAction.ChannelID != channelID {
		channelAction.ChannelID = channelID
	}
```

---

Whoops, didn't see that. Let me change it, thanks again!

---

Heh, changing this surfaced a bug in the tests, so I made *my* life easier :joy: Fixed in https://github.com/mattermost/mattermost-plugin-playbooks/pull/1036/commits/ba08567bc8ef96f4717b076fe70a66eaf991f680 Thanks :)